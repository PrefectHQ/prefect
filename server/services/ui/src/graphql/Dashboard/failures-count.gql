query FailureCount($heartbeat: timestamptz) {
  flow_aggregate(
    where: {
      flow_runs: { state: { _eq: "Failed" }, heartbeat: { _gte: $heartbeat } }
    }
  ) {
    aggregate {
      count
    }
  }
}
