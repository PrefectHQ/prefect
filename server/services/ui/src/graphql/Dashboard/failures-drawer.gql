query FailDetails($heartbeat: timestamptz) {
  flow_run(
    where: { state: { _eq: "Failed" }, heartbeat: { _gte: $heartbeat } }
    distinct_on: flow_id
    order_by: [{ flow_id: asc }, { heartbeat: desc }]
  ) {
    id
    name
    scheduled_start_time
    start_time
    end_time
    duration
    flow {
      id
      name
      failed_runs_count: flow_runs_aggregate(
        where: { state: { _eq: "Failed" }, heartbeat: { _gte: $heartbeat } }
      ) {
        aggregate {
          count
        }
      }
      runs_count: flow_runs_aggregate(
        where: { heartbeat: { _gte: $heartbeat } }
      ) {
        aggregate {
          count
        }
      }
    }
  }
}
