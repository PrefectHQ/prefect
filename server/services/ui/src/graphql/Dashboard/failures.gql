query Failures($heartbeat: timestamptz) {
  flow_run(
    where: {
      heartbeat: { _gte: $heartbeat }
      logs: { level: { _in: ["CRITICAL", "ERROR"] } }
    }
    distinct_on: flow_id
    order_by: [{ flow_id: asc }, { heartbeat: desc }]
  ) {
    id
    name
    logs(
      limit: 1
      where: {
        level: { _in: ["CRITICAL", "ERROR"] }
        timestamp: { _gte: $heartbeat }
      }
    ) {
      id
      level
      message
      name
      timestamp
    }

    flow {
      id
      name
    }
  }
}
