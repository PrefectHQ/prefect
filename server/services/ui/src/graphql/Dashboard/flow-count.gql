query FlowCount($archived: Boolean, $searchParams: [flow_bool_exp!]) {
  flowCount: flow_aggregate(
    where: { archived: { _eq: $archived }, _or: $searchParams }
  ) {
    aggregate {
      count
    }
  }
}
