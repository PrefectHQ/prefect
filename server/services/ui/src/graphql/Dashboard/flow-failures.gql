query($heartbeat: timestamptz) {
  flow {
    id
    name
    archived
    failed_count: flow_runs_aggregate(
      where: { state: { _eq: "Failed" }, heartbeat: { _gte: $heartbeat } }
    ) {
      aggregate {
        count
      }
    }
    runs_count: flow_runs_aggregate(
      where: { heartbeat: { _gte: $heartbeat } }
    ) {
      aggregate {
        count
      }
    }
  }
}
