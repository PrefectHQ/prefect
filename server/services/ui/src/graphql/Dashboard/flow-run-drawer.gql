query FlowRun($id: uuid!) {
  flow_run_by_pk(id: $id) {
    id
    name
    state
    context
    auto_scheduled
    state_message
    duration
    start_time
    end_time
    flow {
      id
      name
      version
    }
    parameters
    scheduled_start_time
  }
}
