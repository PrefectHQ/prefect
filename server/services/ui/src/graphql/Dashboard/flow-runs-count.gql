query FlowRunsCount($heartbeat: timestamptz) {
  flow_run_aggregate(
    where: { heartbeat: { _gte: $heartbeat }, state: { _neq: "Scheduled" } }
  ) {
    aggregate {
      count
    }
  }
}
