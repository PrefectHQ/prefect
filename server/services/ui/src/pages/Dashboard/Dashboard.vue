<script>
import ErrorsTile from '@/pages/Dashboard/Errors-Tile'
import FlowRunHeartbeatTile from '@/pages/Dashboard/FlowRunHeartbeat-Tile'
import SummaryTile from '@/pages/Dashboard/Summary-Tile'
import FlowTableTile from '@/pages/Dashboard/FlowTable-Tile'
import UpcomingRunsTile from '@/pages/Dashboard/UpcomingRuns-Tile'
import FailuresTile from '@/pages/Dashboard/Failures-Tile'
import SubPageNav from '@/layouts/SubPageNav'
import TileLayout from '@/layouts/TileLayout'
import ApiHealthcheckTile from '@/pages/Dashboard/ApiHealthcheck-Tile'

export default {
  components: {
    ErrorsTile,
    FailuresTile,
    FlowRunHeartbeatTile,
    FlowTableTile,
    SubPageNav,
    SummaryTile,
    TileLayout,
    UpcomingRunsTile,
    ApiHealthcheckTile
  },
  data() {
    return {
      loading: 0,
      previousParams: { flows: { flows: '' }, agents: { agents: '' } },
      tab: this.getTab(),
      tabs: [
        {
          name: 'Overview',
          target: 'overview',
          icon: 'view_quilt'
        },
        {
          name: 'Flows',
          target: 'flows',
          icon: 'timeline'
        }
      ]
    }
  },
  computed: {
    hideOnMobile() {
      return { 'tabs-hidden': this.smAndDown }
    },
    smAndDown() {
      return this.$vuetify?.breakpoint?.smAndDown
    }
  },
  watch: {
    tab(val, prev) {
      let query = {}

      switch (val) {
        case 'flows':
          query = this.previousParams.flows
          break
        default:
          break
      }

      this.previousParams[prev] = this.$route.query
      this.$router.replace({ query }).catch(e => e)
    },
    $route(val) {
      if (Object.keys(val.query).length === 0 && this.tab !== 'overview') {
        this.tab = 'overview'
      }
    }
  },
  methods: {
    getTab() {
      if (!this.$route?.query) return 'overview'
      if ('flows' in this.$route.query) return 'flows'
      return 'overview'
    }
  }
}
</script>

<template>
  <v-sheet color="appBackground">
    <SubPageNav>
      <span slot="page-type">Dashboard</span>
      <span
        slot="page-title"
        style="height: 28px;
        overflow: hidden;"
      >
        <span v-if="loading === 0" style="text-transform: capitalize;">
          {{ tab }}
        </span>
        <span v-else>
          <v-skeleton-loader type="heading" tile></v-skeleton-loader>
        </span>
      </span>
    </SubPageNav>

    <v-tabs
      v-model="tab"
      class="px-6 mx-auto tabs-border-bottom"
      :class="hideOnMobile"
      style="max-width: 1440px;"
      light
    >
      <v-tabs-slider color="blue"></v-tabs-slider>

      <v-tab
        v-for="tb in tabs"
        :key="tb.target"
        :data-cy="`dashboard-${tb.target}-tab`"
        :href="`#${tb.target}`"
        :style="hideOnMobile"
      >
        <v-icon left :size="tb.iconSize || 'medium'">{{ tb.icon }}</v-icon>
        {{ tb.name }}
      </v-tab>

      <v-tab-item class="tab-full-height pa-0" value="overview">
        <TileLayout>
          <SummaryTile slot="row-1-col-1-tile-1" full-height />

          <FailuresTile slot="row-1-col-2-tile-1" />

          <UpcomingRunsTile slot="row-1-col-3-tile-1" full-height />

          <FlowRunHeartbeatTile slot="row-2-col-1-row-2-tile-1" />

          <ErrorsTile slot="row-2-col-2-row-2-tile-1" full-height />

          <ApiHealthcheckTile slot="row-2-col-2-row-2-tile-2" />
        </TileLayout>
      </v-tab-item>

      <v-tab-item class="tab-full-height" value="flows">
        <FlowTableTile v-if="tab == 'flows'" class="mx-3 my-6" />
      </v-tab-item>

      <v-tab-item class="tab-full-height" value="analytics">
        Nothing to see here :)
      </v-tab-item>
    </v-tabs>

    <v-bottom-navigation v-if="smAndDown" fixed>
      <v-btn v-for="tb in tabs" :key="tb.target" @click="tab = tb.target">
        {{ tb.name }}
        <v-icon>{{ tb.icon }}</v-icon>
      </v-btn>
    </v-bottom-navigation>
  </v-sheet>
</template>

<style lang="scss">
.tab-full-height {
  min-height: 100vh;
}
</style>
