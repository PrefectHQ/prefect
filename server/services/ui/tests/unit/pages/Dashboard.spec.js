// IMPORTANT: Do not remove this import since it mocks all the graphql routes used on the dashboard
// eslint-disable-next-line no-unused-vars
import graphqlMocks from './graphqlMocks'

// Libraries
import Vuetify from 'vuetify'
import Vue from 'vue'

// Components
import Dashboard from '@/pages/Dashboard/Dashboard'

// Utilities
import { createLocalVue, shallowMount } from '@vue/test-utils'

const localVue = createLocalVue()
Vue.use(Vuetify)

describe('Dashboard', () => {
  let vuetify

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  let wrapper = shallowMount(Dashboard, {
    localVue,
    vuetify
  })

  it('renders a vue instance', () => {
    expect(wrapper.isVueInstance()).toBe(true)
  })

  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should properly mount', () => {
    const wrapper = shallowMount(Dashboard, {})

    expect(wrapper.element).toMatchSnapshot()
  })
})
