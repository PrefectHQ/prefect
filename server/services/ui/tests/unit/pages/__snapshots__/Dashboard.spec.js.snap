// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should match snapshot 1`] = `
<v-sheet-stub
  color="appBackground"
  tag="div"
>
  <subpagenav-stub>
    <span>
      Dashboard
    </span>
     
    <span
      style="height: 28px; overflow: hidden;"
    >
      <span
        style="text-transform: capitalize;"
      >
        
        overview
      
      </span>
    </span>
  </subpagenav-stub>
   
  <v-tabs-stub
    activeclass=""
    class="px-6 mx-auto tabs-border-bottom"
    light="true"
    mobilebreakpoint="1264"
    nexticon="$next"
    previcon="$prev"
    slidersize="2"
    style="max-width: 1440px;"
    value="overview"
  >
    <v-tabs-slider-stub
      color="blue"
    />
     
    <v-tab-stub
      data-cy="dashboard-overview-tab"
      href="#overview"
      ripple="true"
    >
      <v-icon-stub
        left=""
        size="medium"
      >
        view_quilt
      </v-icon-stub>
      
      Overview
    
    </v-tab-stub>
    <v-tab-stub
      data-cy="dashboard-flows-tab"
      href="#flows"
      ripple="true"
    >
      <v-icon-stub
        left=""
        size="medium"
      >
        timeline
      </v-icon-stub>
      
      Flows
    
    </v-tab-stub>
     
    <v-tab-item-stub
      class="tab-full-height pa-0"
      value="overview"
    >
      <tilelayout-stub>
        <summarytile-stub
          fullheight="true"
        />
         
        <failurestile-stub />
         
        <upcomingrunstile-stub
          fullheight="true"
        />
         
        <flowrunheartbeattile-stub />
         
        <errorstile-stub
          fullheight="true"
        />
         
        <apihealthchecktile-stub />
      </tilelayout-stub>
    </v-tab-item-stub>
     
    <v-tab-item-stub
      class="tab-full-height"
      value="flows"
    >
      <!---->
    </v-tab-item-stub>
     
    <v-tab-item-stub
      class="tab-full-height"
      value="analytics"
    >
      
      Nothing to see here :)
    
    </v-tab-item-stub>
  </v-tabs-stub>
   
  <!---->
</v-sheet-stub>
`;

exports[`Dashboard should properly mount 1`] = `
<v-sheet-stub
  color="appBackground"
  tag="div"
>
  <subpagenav-stub>
    <span>
      Dashboard
    </span>
     
    <span
      style="height: 28px; overflow: hidden;"
    >
      <span
        style="text-transform: capitalize;"
      >
        
        overview
      
      </span>
    </span>
  </subpagenav-stub>
   
  <v-tabs-stub
    activeclass=""
    class="px-6 mx-auto tabs-border-bottom"
    light="true"
    mobilebreakpoint="1264"
    nexticon="$next"
    previcon="$prev"
    slidersize="2"
    style="max-width: 1440px;"
    value="overview"
  >
    <v-tabs-slider-stub
      color="blue"
    />
     
    <v-tab-stub
      data-cy="dashboard-overview-tab"
      href="#overview"
      ripple="true"
    >
      <v-icon-stub
        left=""
        size="medium"
      >
        view_quilt
      </v-icon-stub>
      
      Overview
    
    </v-tab-stub>
    <v-tab-stub
      data-cy="dashboard-flows-tab"
      href="#flows"
      ripple="true"
    >
      <v-icon-stub
        left=""
        size="medium"
      >
        timeline
      </v-icon-stub>
      
      Flows
    
    </v-tab-stub>
     
    <v-tab-item-stub
      class="tab-full-height pa-0"
      value="overview"
    >
      <tilelayout-stub>
        <summarytile-stub
          fullheight="true"
        />
         
        <failurestile-stub />
         
        <upcomingrunstile-stub
          fullheight="true"
        />
         
        <flowrunheartbeattile-stub />
         
        <errorstile-stub
          fullheight="true"
        />
         
        <apihealthchecktile-stub />
      </tilelayout-stub>
    </v-tab-item-stub>
     
    <v-tab-item-stub
      class="tab-full-height"
      value="flows"
    >
      <!---->
    </v-tab-item-stub>
     
    <v-tab-item-stub
      class="tab-full-height"
      value="analytics"
    >
      
      Nothing to see here :)
    
    </v-tab-item-stub>
  </v-tabs-stub>
   
  <!---->
</v-sheet-stub>
`;
